decoder.go

```diff
@@ -454,39 +456,79 @@ func (d *decoder) pointer(name string, value reflect.Value, def string) error {
 	return d.unmarshal(name, value.Elem(), def)
 }
 
-// Sets value to a channel type.
-func (d *decoder) channel(key string, value reflect.Value, def string) error {
-	return d.textUnmarshaller(key, value, def)
-}
-
-// Sets value to a function type.
-func (d *decoder) function(key string, value reflect.Value, def string) error {
-	return d.textUnmarshaller(key, value, def)
-}
+// JSON encoder will fail to serialize maps that don't have strings as keys
+// so we are going to stringify them manually.
+func jsonMap(v interface{}) interface{} {
+	if reflect.TypeOf(v).Kind() == reflect.Map {
+		tmp := make(map[string]interface{})
+		rv := reflect.ValueOf(v)
+		for _, key := range rv.MapKeys() {
+			tmp[fmt.Sprint(key.Interface())] = jsonMap(rv.MapIndex(key).Interface())
+		}
 
-func (d *decoder) textUnmarshaller(key string, value reflect.Value, str string) error {
-	v := d.getGlobalProvider().Get(key)
-	if v.HasValue() {
-		str = v.String()
-	} else if str == "" {
-		return nil
+		return tmp
 	}
 
-	if value.IsNil() {
-		value.Set(reflect.New(value.Type()).Elem())
-	}
+	return v
+}
+
+// tryUnmarshallers checks if the value's type implements either one of standard interfaces in order:
+// 1. `json.Unmarshaler`
+// 2. `encoding.TextUnmarshaler`
+// 3. `yaml.Unmarshaler`
+// and tries it to populate the value.
+func (d *decoder) tryUnmarshalers(key string, value reflect.Value, def string) (bool, error) {
+	switch value.Kind() {
+	// value.IsNil panics if value.Kind() is not equal to one of these constants.
+	case reflect.Chan, reflect.Interface, reflect.Func, reflect.Map, reflect.Ptr:
+		if value.IsNil() {
+			value.Set(reflect.New(value.Type()).Elem())
+		}
 
-	// Value has to have a pointer receiver to be able to modify itself with TextUnmarshaller
-	if !value.CanAddr() {
-		return errorWithKey(errors.New("can't use TextUnmarshaller because value is not addressable"), key)
+	//TODO(alsam) Command-line provider implementation is made via maps, so we'll fail to merge slices and maps
+	// TestCommandLineProvider_NestedValues will fail in that case.
+	case reflect.Slice:
```

Comment:
Command line provider implements slices as maps internally.
When user asks to populate e.g. `roles` slice it will return a map and group provider will fail to merge it with an array. Decoder's slice method skips request for `roles` key and starts populating elements from the very beginning: `roles.0`, `roles.1` ...

Let me create an issue for this TODO.